cura/Machines/ContainerNode.py

```diff
@@ -0,0 +1,46 @@
+from typing import Optional
+
+from collections import OrderedDict
+
+from UM.Logger import Logger
+from UM.Settings.InstanceContainer import InstanceContainer
+
+
+##
+# A metadata / container combination. Use getContainer() to get the container corresponding to the metadata.
+#
+# ContainerNode is a multi-purpose class. It has two main purposes:
+#  1. It encapsulates an InstanceContainer. It contains that InstanceContainer's
+#          - metadata (Always)
+#          - container (lazy-loaded when needed)
+#  2. It also serves as a node in a hierarchical InstanceContainer lookup table/tree.
+#     This is used in Variant, Material, and Quality Managers.
```

Comment:
A container already has metadata. I have two problems with this approach, one for performance and the other for architecture:

1. The `findContainersMetadata` function was explicitly chosen to return a `dict` because the `dict` is fast to construct. One major problem with the original start-up sequence was that it needs to construct a lot of Python objects. Grouping the metadata and an optional container together in a class requires creating these instances again on start-up.
2. The metadata is required in the node as well as in the container. The wrapper is unnecessary then (since the data can be requested from the container just as well) and it is less confusing to make the setting data in the container optional instead. Then you won't need a wrapper class and when using the nodes you won't need to request the container of a container.

These two concerns don't fit together. I'd say that the second concern is more important and that a container should be made lazy by itself when requesting its setting data.

In short, I'm proposing to merge this wrapper class in with the actual container in Uranium.

Reply:
After some discussion, we agree that the concept of this `ContainerNode` can be moved into Uranium and many things can benefit from it. This doesn't seem to be a small work, so @jackha and I think it's better not to do it right now. For reference, we have created a ticket for this: CURA-5027